numba/array_analysis.py

```diff
@@ -41,846 +52,1266 @@
+        """Try to derive if given objects are equivalence, return true
+        if so, or false otherwise.
+        """
+        inds = [self._get_ind(x) for x in objs]
+        ind = max(inds)
+        if ind != -1:
+            return all(i == ind for i in inds)
+        else:
+            return all([x == objs[0] for x in objs])
+
+    def get_equiv_const(self, obj):
+        """Check if obj is equivalent to some int constant, and return
+        the constant if found, or None otherwise.
+        """
+        ind = self._get_ind(obj)
+        if ind >= 0:
+            objs = self.ind_to_obj[ind]
+            for x in objs:
+                if isinstance(x, int):
+                    return x
+        return None
+
+    def get_equiv_set(self, obj):
+        """Return the set of equivalent objects.
+        """
+        ind = self._get_ind(obj)
+        if ind >= 0:
+            return set(self.ind_to_obj[ind])
+        return set()
+
+    def insert_equiv(self, *objs):
+        """Insert a set of equivalent objects by modifying self. This
+        method can be overloaded to transform object type before insertion.
+        """
+        self._insert(objs)
+
+    def intersect(self, equiv_set):
+        """ Return the intersection of self and the given equiv_set,
+        without modifying either of them. The result will also keep
+        old equivalence indices unchanged.
+        """
+        new_set = self.empty()
+        new_set.next_ind = self.next_ind
+
+        for objs in equiv_set.ind_to_obj.values():
+            inds = tuple(self._get_ind(x) for x in objs)
+            ind_to_obj = {}
+
+            for i, x in zip(inds, objs):
+                if i in ind_to_obj:
+                    ind_to_obj[i].append(x)
+                elif i >= 0:
+                    ind_to_obj[i] = [x]
+
+            for k, v in ind_to_obj.items():
+                if len(v) > 1:
+                    new_set._insert(v)
+
+        return new_set
+
+
+class ShapeEquivSet(EquivSet):
+    """Just like EquivSet, except that it accepts only numba IR variables
+    and constants as objects, guided by their types. Arrays are considered
+    equivalence as long as their shapes are equivalent. Scalars are
+    equivalent only when they are equal in value. Tuples are equivalent
+    when they are of the same size, and their elements are equivalent.
+    """
+    def __init__(self, typemap, defs, ind_to_var = None, obj_to_ind = None,
+                 ind_to_obj = None, next_id = 0):
+        self.typemap = typemap
+        self.defs = defs
+        self.ind_to_var = ind_to_var if ind_to_var else {}
+        super(ShapeEquivSet, self).__init__(obj_to_ind, ind_to_obj, next_id)
+
+    def empty(self):
+        """Return an empty ShapeEquivSet.
+        """
+        return ShapeEquivSet(self.typemap, {})
+
+    def clone(self):
+        """Return a new copy.
+        """
+        return ShapeEquivSet(
+                   self.typemap, copy.copy(self.defs),
```

Comment:
No, `typemap` is shared, and this is safe because it is not modified by the ShapeEquivSet object.